File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_58_copy_list_with_random_pointer.ts
import { _Node } from "./Node";

function copyRandomList(head: _Node | null): _Node | null {
    /**
     * Approach: Using HashMap
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     */
    if (head === null) {
        return null;
    }

    // Map old nodes to new nodes
    const oldToNew = new Map<_Node, _Node>();

    // First pass: create all new nodes
    let curr: _Node | null = head;
    while (curr !== null) {
        oldToNew.set(curr, new _Node(curr.val));
        curr = curr.next;
    }

    // Second pass: assign next and random pointers
    curr = head;
    while (curr !== null) {
        const newNode = oldToNew.get(curr)!;

        if (curr.next !== null) {
            newNode.next = oldToNew.get(curr.next)!;
        }

        if (curr.random !== null) {
            newNode.random = oldToNew.get(curr.random)!;
        }

        curr = curr.next;
    }

    return oldToNew.get(head)!;
}

File: src/my_project/interviews/top_150_questions_round_22/ex_58_copy_list_with_random_pointer.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod

# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        """
        Approach 1: Using HashMap
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        if not head:
            return None
        
        # Map old nodes to new nodes
        old_to_new = {}
        
        # First pass: create all new nodes
        curr = head
        while curr:
            old_to_new[curr] = Node(curr.val)
            curr = curr.next
        
        # Second pass: assign next and random pointers
        curr = head
        while curr:
            if curr.next:
                old_to_new[curr].next = old_to_new[curr.next]
            if curr.random:
                old_to_new[curr].random = old_to_new[curr.random]
            curr = curr.next
        
        return old_to_new[head]
